### ISSUE TITLE
Preventing offline behaviour on mutations

### ISSUE THREAD

**Issue by kirkness** [Malpaux/apollo-offline#6]
I am wondering whether there is a way of preventing the offline behaviour on mutations. It seems to just to not resolve/reject when calling them.

**Comment by PaulBrachmann** [comment]
Hey @kirkness,
I don't quite understand what you mean by

> It seems to just to not resolve/reject when calling them.

Could you please clarify where in the public API you expect promises to be returned? - Or have you noticed something not working properly internally?

You want your mutations to be executed once and then discarded if they fail, right?

If so, there currently is no way to selectively exclude certain mutations/queries from being retried. However, I'd be happy to add this feature in the next release.

Have a great evening!

**Comment by kirkness** [comment]
Hey @MLPXBrachmann, thanks for getting back - sorry if I wasn't clear. Currently if I call a mutation like so:

```js
try {
  this.setState({ isLoading: true });
  const data = await props.mutation();
  Alert.alert('Success');
} catch(e) {
  Alert.alert('Error');
} finally {
  this.setState({ isLoading: false });
}
```

Neither of the alerts will be called and it will remain loading, is this intended behaviour or am I doing something wrong? Ideally, as I say, id like to prevent the user from attempting mutations offline and just show a network error - so preventing the offline cache on select queries/mutations would be a very nice feature!

p.s. Thanks a lot for putting this together, I had the classic "Oh and the app should work offline" on the day of release - and this dramatically reduced my stress! 👍

**Comment by PaulBrachmann** [comment]
@kirkness, thank you very much for the kind words!

Assuming this occurred while the app couldn't reach your backend: Yes, this is the intended behaviour.  
Apollo-Offline does not resolve mutations until they've been delivered to the server or an [irreconcilable error](https://github.com/jevakallio/redux-offline#change-how-irreconcilable-errors-are-detected) occurred. This is one of the cornerstones of the offline-first approach since it allows users to work offline and guarantees that all changes will be synced once they go back online. Before they do, there is now way to tell whether a mutation will ultimately succeed or fail.

I can, however, see the benefits of being able to selectively disable this behaviour. Expect that feature to be implemented by tomorrow.

It is likely going to be enabled by adding a ```__online__``` field with a truthy value to your query/mutation variables. Doing this will allow you to revert back to Apollo's default behaviour for that specific query/mutation.

I hope I understood you correctly and the proposed feature will meet your requirements.  
If not - or if you have any additional questions - feel free to just comment again!

**Comment by kirkness** [comment]
Legend - thanks for clearing that up for me! `__online__` is exactly what I'm after.

**Comment by PaulBrachmann** [comment]
I added selective opt-out through the ```__online__``` flag in b8f928a. It is available from v0.0.6 onwards (you may need to upgrade manually).

Thank you for helping to improve Apollo-Offline!

**Comment by PaulBrachmann** [comment]
@kirkness May I close this issue?
